### ISSUE TITLE
Does gradient checkpoint works in these codes?

### ISSUE THREAD

**Issue by ewrfcas** [allenai/longformer#53]
I have seen gradient checkpoint in the paper but not in this git?

**Comment by ibeltagy** [comment]
Good point. Implementing gradient checkpointing requires a small change in `BertEncoder`, which I added here https://github.com/huggingface/transformers/pull/4659, but this PR is unlikely to be merged soon. In the meantime, you can use my branch [here](https://github.com/ibeltagy/transformers/tree/grad_checkpointing).

**Comment by ewrfcas** [comment]
Thanks for your response! 
And I have another question that when I am using the demo with length=4096, batchsize=1, the mask is sliding chunks with (0,1), fp16 in tvm gpu, it failed with OOM in Tesla-V100 (16GB). Does it work properly? If so, I think gradient checkpoint maybe necessary for training, because even batchsize=1 can't be inferenced successfully in fp16 16GB GPU.

**Comment by ewrfcas** [comment]
The demo is 
```import torch
from longformer.longformer import Longformer, LongformerConfig
from longformer.sliding_chunks import pad_to_window_size
from transformers import RobertaTokenizer

config = LongformerConfig.from_pretrained('./longformer-large-4096/')
# choose the attention mode 'n2', 'tvm' or 'sliding_chunks'
# 'n2': for regular n2 attantion
# 'tvm': a custom CUDA kernel implementation of our sliding window attention
# 'sliding_chunks': a PyTorch implementation of our sliding window attention
config.attention_mode = 'tvm'

model = Longformer.from_pretrained('./longformer-large-4096/', config=config)
tokenizer = RobertaTokenizer(vocab_file='/users/caochenjie/DEFT/roberta_large/vocab.json',
                             merges_file='/users/caochenjie/DEFT/roberta_large/merges.txt')
tokenizer.model_max_length = model.config.max_position_embeddings

SAMPLE_TEXT = ' '.join(['Hello world! '] * 1000)  # long input document

input_ids = torch.tensor(tokenizer.encode(SAMPLE_TEXT)).unsqueeze(0)  # batch of size 1
model = model.half()
# TVM code doesn't work on CPU.
# Uncomment this if `config.attention_mode = 'tvm'`
model = model.cuda()
input_ids = input_ids.cuda()

# Attention mask values -- 0: no attention, 1: local attention, 2: global attention
attention_mask = torch.ones(input_ids.shape, dtype=torch.long, device=input_ids.device)  # initialize to local attention
# attention_mask[:, [1, 4, 21, ]] = 2  # Set global attention based on the task. For example,
# classification: the <s> token
# QA: question tokens

# padding seqlen to the nearest multiple of 512. Needed for the 'sliding_chunks' attention
input_ids, attention_mask = pad_to_window_size(
    input_ids, attention_mask, config.attention_window[0], tokenizer.pad_token_id)
print(input_ids.shape, attention_mask.shape)
output = model(input_ids, attention_mask=attention_mask)[0]
print(output)
```

**Comment by ibeltagy** [comment]
I see you are using `model.half()`. You also need `input_ids.half()` and `attention_mask.half()`. That said, once you start adding some global attention, the 16GB memory won't be enough unless you use gradient checkpointing.

**Comment by ewrfcas** [comment]
input_ids will be changed into embedding with fp16, and attention_mask will cast into fp16 too in transformers, so it should works properly. Thanks for your advices, I will try the gradient checkpoint in https://github.com/ibeltagy/transformers/tree/grad_checkpointing. 
Can you offer the MLM training batchsize for each 48GB gpu during the pretrain? It will help me a lot for testing the gradient checkpoint during the pretraining, thanks very much!

**Comment by ibeltagy** [comment]
> Can you offer the MLM training batchsize for each 48GB gpu during the pretrain?

If I remember correctly, it was 2 but the memory wasn't full, and this was without gradient checkpointing.

**Comment by ewrfcas** [comment]
Thanks for your advice!
